manual/src/tutorials/memorymodel.etex

```diff
@@ -0,0 +1,783 @@
+free is made, they do not need to worry about reorderings in their code.
+
+\section{s:drf_reasoning}{Reasoning with DRF-SC}
+
+In this section, we will look at examples of using DRF-SC for program
+reasoning. In this section, we will use the functions with names "dN" to
+represent domains executing in parallel with other domains.
+
+Here is a simple example with a data race:
+
+\begin{caml_example*}{verbatim}
+(* Has data race *)
+let r = ref 0
+let d1 () = r := 1
+let d2 () = !r
+\end{caml_example*}
+
+"r" is a non-atomic reference. The two domains race to access the reference,
+and "d1" is a write. Since there is no happens-before relationship between the
+conflicting accesses, there is a data race.
+
+Both of the programs that we had seen in the section~\ref{s:why_relaxed_memory}
+have data races. It is no surprise that they exhibit non sequentially
+consistent behaviours.
+
+Accessing disjoint array indices and fields of a record in parallel is not a
+data race. For example,
+
+\begin{caml_example*}{verbatim}
+(* No data race *)
+let a = [| 0; 1 |]
+let d1 () = a.(0) <- 42
+let d2 () = a.(1) <- 42
+\end{caml_example*}
+
+\begin{caml_example*}{verbatim}
+(* No data race *)
+type t = {
+  mutable a : int;
+  mutable b : int
+}
+let r = {a = 0; b = 1}
+let d1 () = r.a <- 42
+let d2 () = r.b <- 42
+\end{caml_example*}
+
+\noindent do not have data races.
+
+Races on atomic locations do not lead to a data race.
+
+\begin{caml_example*}{verbatim}
+(* No data race *)
+let r = Atomic.make 0
+let d1 () = Atomic.set r 1
+let d2 () = Atomic.get r
+\end{caml_example*}
+
+\subsubsection{s:mm_msg_passing}{Message-passing}
+
+Atomic variables may be used for implementing non-blocking communication
+between domains.
+
+\begin{caml_example*}{verbatim}
+(* No data race *)
+let msg = ref 0
+let flag = Atomic.make false
+let d1 () =
+  msg := 42; (* a *)
+  Atomic.set flag true (* b *)
+let d2 () =
+  if Atomic.get flag (* c *) then
+    !msg (* d *)
+  else 0
+\end{caml_example*}
+
+Observe that the actions "a" and "d" write and read from the same non-atomic
+location "msg", respectively, and hence are conflicting. We need to establish
+that "a" and "d" have a happens-before relationship in order to show that this
+program does not have a data race.
+
+The action "a" precedes "b" in program order, and hence, "a" happens-before
+"b". Similarly, "c" happens-before "d". If "d2" observes the atomic variable
+"flag" to be "true", then "b" precedes "c" in happens-before order. Since
+happens-before is transitive, the conflicting actions "a" and "d" are in
+happens-before order. If "d2" observes the "flag" to be "false", then the read
+of "msg" is not done. Hence, there is no conflicting access in this execution
+trace. Hence, the program does not have a data race.
+
+The following modified version of the message passing program does have a data
+race.
+
+\begin{caml_example*}{verbatim}
+(* No data race *)
```

Comment:
Shouldn't this comment indicate there is a date race in this example?

Reply:
Good spot. Fixed. Thanks.